ring_buffer: Introduce `RING_BUF_ITEM_SIZEOF`

hx-bora opened the pull request:
Current SIZE32_OF only works on variables, not types, due to an extra parenthesis pair.
Indeed, sizeof((int)) is not valid C, whereas sizeof((my_var)) is.

Remove the extra parenthesis pair.

andyross reviewed (changes requested):
This doesn't look like a public API to me.  It's just some random thing in the ring buffer header that probably should have a Z_ prefix on it.

If the ring buffer code doesn't pass a type, then why do we care?  If this is something you want to use in an app, I'd suggest moving this into sys/util.h, documenting it, ... and then having the fight about whether a "size in 32 bit words" is a useful API to have or not.

andyross commented:
Actually thinking harder, I'd put a -1 on this as a public API too.  It **rounds down**.  So SIZE32_OF(whatever) is going to be too small for anything that isn't correctly sized!  That's just dangerous.

cataldor commented:
Hello andyross,
I'm not sure what counts as public API or not for zephyr, but it is documented in zephyr API [1] and is explicitly mentioned as "user-accessible" (as long as it is placed in a proper position)

This macro is supposed to be used when employing the "item-based" ring buffer: in this case, all items must be 32-bits (so, there is a type). Therefore, the rounding-down cannot happen, as long as the user respects the API.

It seems to me that there are three ways to allocate a ring-buffer for item-based:
1. Use a uint32_t buffer and use the `SIZE32_OF` macro, and use `SIZE32_OF` for the item-based API (ring buffer size);
2. Use `RING_BUF_ITEM_DECLARE` (but this will not work inside a struct, for instance); and
3. Use a uint8_t buffer and `sizeof`, but you will need to take care with the API (i.e., you cannot pass `sizeof` for item-based API calls)

I would say the `SIZE32_OF` is useful if you plan to use the item-based ring buffer. However, AFAIK, there are only tests on the tree using this API.

[1] https://docs.zephyrproject.org/latest/kernel/data_structures/ring_buffers.html

andyross commented:
This macro was never documented.  And it's actually not used anywhere at all by the ringbuffer code, only in a few spots in tests/lib/ringbuffer.  The macro as-is should be moved into `tests/lib/ringbuffer/src/main.c`, which is the only file that uses it.

Obviously this is a minor thing, but I think the -1 stays.  I just don't like this API.

If you do want a public API for this, then define and document something like a RING_BUF_ITEM_SIZEOF() macro which includes an assertion check for the round-down problem.

cataldor commented:
Ah, you were talking just about the macro. I see.

andyross commented:
Yeah, good point.  Rounding up seems fine to me.  The overwhelming majority of cases will be called on static/compile-time-visible data anyway and the extra math will drop out.

Commit:
ring_buffer: Introduce `RING_BUF_ITEM_SIZEOF`

Currently, to compute the 'item' size in a ring buffer, we have
`SIZE32_OF`.

Several issues with this:
- `SIZE32_OF` only works on variables, not types, due to an extra
 parenthesis pair. Indeed, `sizeof((int))` is not valid C, whereas
 `sizeof((my_var))` is.
- `SIZE32_OF` is not a proper public API
- `SIZE32_OF` rounds down if the argument size is not a multiple
 of 4 bytes.

Thus, we introduce a proper `RING_BUF_ITEM_SIZEOF`, fixing the
aforementioned issues.

Signed-off-by: Henri Xavier <datacomos@huawei.com>

andyross reviewed (approved):
Sorry, this fell off the radar.  Looks fine to me.
